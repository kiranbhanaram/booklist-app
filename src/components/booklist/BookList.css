.bookListContainer {
    overflow-y: auto;
    height: 80vh;
    line-height: 1.6;
    -webkit-font-smoothing: antialiased;
    color: #fff;
    font-size: 1rem;
    background-color: #2C3845;
    margin: 20px;
  }
  .bookListContainer thead th {
    position: sticky;
    top: 0;
  }
  .bookListContainer table {
    border-collapse: collapse;
    width: 100%;
  }
  .bookListContainer table th,
  .bookListContainer table td {
    padding: 8px 16px;
    border: 1px solid #ccc;
  }
  .bookListContainer table th {
    width: 100%;
    background-color: #202932;
  }
  
  tbody tr:hover{
  background:#BDC3C7;
    color:#FFFFFF;
    cursor: pointer;
  }