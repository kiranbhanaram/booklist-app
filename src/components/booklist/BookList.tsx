import { useEffect, useState } from 'react';
import { axiosInstance } from '../../axiosConfig';
import { Loader } from '../loader/Loader';
import { Toast } from '../toast/Toast';
import './BookList.css';
import { BookDetail } from '../bookDetail/BookDetail';

type ErrorProp = {
  message: string;
};

export type BookProps = {
  name?: string;
  author?: string;
  language: string;
  direction: string;
  source: string;
  comments: string;
};

export function BookList() {
  const [tableData, setTableData] = useState([]);
  const [showLoader, setShowLoader] = useState(false);
  const [errorMsg, setErrorMsg] = useState('');
  const [modalData, setModalData] = useState<BookProps>();

  useEffect(() => {
    getTableData();
  }, []);

  const getTableData = (): void => {
    try {
      setShowLoader(true);
      axiosInstance
        .get('fawazahmed0/quran-api@1/editions.json')
        .then((response) => {
          if (response.data) {
            if (Object.values(response.data).length > 0) {
              setTableData(Object.values(response.data));
            }
          }
          setShowLoader(false);
        })
        .catch((error) => {
          setTableData([]);
          handleError(error, 'Table Data');
        });
    } catch (error) {
      setShowLoader(false);
    }
  };

  const handleError = (error: ErrorProp, apiName: string): void => {
    setShowLoader(false);
    setErrorMsg(error.message || 'Error while making request to server');
    setTimeout(() => setErrorMsg(''), 3000);
  };

  return (
    <>
      {showLoader ? (
        <Loader />
      ) : (
        <div className="bookListContainer">
          <Toast message={errorMsg} />
          <table>
            <thead>
              <tr>
                <th style={{ width: '49%' }}>Book Name</th>
                <th>Author</th>
              </tr>
            </thead>

            <tbody>
              {tableData.map((rowData: BookProps, index: number) => (
                <tr key={index} onClick={() => setModalData(rowData)}>
                  <td>{rowData.name}</td>
                  <td>{rowData.author}</td>
                </tr>
              ))}
            </tbody>
          </table>
          {modalData && <BookDetail rowData={modalData} onClose={() => setModalData(undefined)} />}
        </div>
      )}
    </>
  );
}
