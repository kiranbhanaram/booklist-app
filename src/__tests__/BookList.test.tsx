import { shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { BookList } from '../components/booklist/BookList';

it('renders BookList component', () => {
  expect(
    shallow(
      <MemoryRouter initialEntries={['/']}>
        <BookList />
      </MemoryRouter>
    )
  ).toMatchSnapshot();
});
