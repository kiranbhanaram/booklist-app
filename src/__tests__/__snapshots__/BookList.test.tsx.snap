// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BookList component 1`] = `ShallowWrapper {}`;
